shared-libs/cht-datasource/src/local/contact.ts

```diff
@@ -0,0 +1,179 @@
+        return contact;
+      }
+
+      const combinedContacts: NonEmptyArray<Nullable<Doc>> = [contact, ...lineageContacts];
+      const contactUuids = getPrimaryContactIds(combinedContacts);
+      const contacts = await getMedicDocsById(contactUuids);
+      const [contactWithContact, ...lineageContactsWithContact] = combinedContacts.map(
+        hydratePrimaryContact(contacts)
+      ).filter(item => item ?? false);
+      const contactWithLineage = hydrateLineage(
+        contactWithContact as ContactType.v1.Contact,
+        lineageContactsWithContact
+      );
+      return deepCopy(contactWithLineage);
+    };
+  };
+
+  /** @internal */
+  export const getPage = ({ medicDb, settings }: LocalDataContext) => {
+    // Define query functions
+    const getContactsByTypeFreetext = queryDocsByKey(medicDb, 'medic-client/contacts_by_type_freetext');
+    const getContactsByFreetext = queryDocsByKey(medicDb, 'medic-client/contacts_by_freetext');
+    const getContactsByType = queryDocsByKey(medicDb, 'medic-client/contacts_by_type');
+    const getContactsByTypeFreetextRange = queryDocsByRange(medicDb, 'medic-client/contacts_by_type_freetext');
+    const getContactsByFreetextRange = queryDocsByRange(medicDb, 'medic-client/contacts_by_freetext');
+
+    const determineGetDocsFn = (
+      qualifier: ContactTypeQualifier | FreetextQualifier
+    ): ((limit: number, skip: number) => Promise<Nullable<Doc>[]>) => {
+      if (isContactTypeAndFreetextType(qualifier)) {
+        return getDocsFnForContactTypeAndFreetext(qualifier);
+      }
+
+      if (ContactType.v1.isContactType(qualifier)) {
+        return getDocsFnForContactType(qualifier);
+      }
+
+      if (ContactType.v1.isFreetextType(qualifier)) {
+        return getDocsFnForFreetextType(qualifier);
+      }
+
+      // This code should never be reached due to the type system,
+      // but TypeScript requires it for exhaustiveness checking
+      throw new InvalidArgumentError(`Unsupported qualifier type: ${JSON.stringify(qualifier)}`);
+    };
+
+    const isContactTypeAndFreetextType = (
+      qualifier: ContactTypeQualifier | FreetextQualifier
+    ): qualifier is ContactTypeQualifier & FreetextQualifier => {
+      return ContactType.v1.isContactType(qualifier) && ContactType.v1.isFreetextType(qualifier);
+    };
+
+    const getDocsFnForContactTypeAndFreetext = (
+      qualifier: ContactTypeQualifier & FreetextQualifier
+    ): (limit: number, skip: number) => Promise<Nullable<Doc>[]> => {
+      if (qualifier.freetext.includes(':')) {
+        return (limit, skip) => getContactsByTypeFreetext(
+          [qualifier.contactType, qualifier.freetext],
+          limit,
+          skip
+        );
+      }
+
+      return (limit, skip) => getContactsByTypeFreetextRange(
+        [qualifier.contactType, qualifier.freetext],
+        [qualifier.contactType, qualifier.freetext + END_OF_ALPHABET],
```

Comment:
In this case, the range seems repetitive but the range could be anything from eg `a` to `b` or even `a` to `a`, in those cases, this function will provide flexibility to the user. Another case is [here](https://github.com/medic/cht-core/blob/cb4cce330a99915fc9100ebb8156f201ff991225/shared-libs/cht-datasource/src/local/libs/lineage.ts#L21:L24), where both the start and end key are almost the same. So, applying the range within `getContactsByTypeFreetextRange` would lead to inflexibility.

Reply:
Do we implement support for custom ranges? 
These are generally quite strange and we don't use them anywhere - because of risking to have a lot of trash results that can potentially impact the performance of the query.

Reply:
Not to my knowledge but it's nice to have the option.